### PR TITLE
Document autoLogin configuration #171

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

File: docs/index.adoc [hunk 1]
```diff
@@ -45,15 +45,17 @@ It has the following format:  `[VHost source]/_/idprovider/[ID Provider name]`.
 
 You will need to gather the following information from the Authorization Server setup, for the <<#id_provider_creation, ID Provider configuration>>
 
-. The Client ID
-. The Client secret
+. The Client ID. Must be set, if log in via form is enabled.
```
Comment (rymsha) [inline docs/index.adoc:48]:
```suggestion
. The Client ID. Not required authentication is done only by Auto Login.
```

File: docs/index.adoc [hunk 2]
```diff
@@ -45,15 +45,17 @@ It has the following format:  `[VHost source]/_/idprovider/[ID Provider name]`.
...
+. The Client secret. Must be set, if log in via form is enabled.
```
Comment (rymsha) [inline docs/index.adoc:49]:
```suggestion
. The Client secret. Not required authentication is done only by Auto Login.
```

File: docs/index.adoc [hunk 3]
```diff
@@ -45,15 +45,17 @@ It has the following format:  `[VHost source]/_/idprovider/[ID Provider name]`.
...
 . The Issuer Identifier (URL)
 . The Authorization endpoint URL
 . The Token endpoint URL
+. The JWKS URI
+. The `oidcWellKnownEndpoint` (OpenID Provider Configuration URL), to obtain the values for the above fields and for `userinfoUrl`.
 
 You might have been given only the Issuer URL instead of the list of endpoint URLs.
 In that case the endpoint URLs are accessible under: `[Issuer URL]/.well-known/openid-configuration`
-The JSON result will contain the required information: `issuer`, `authorization_endpoint` and `token_endpoint`.
+The JSON result will contain the required information: `issuer`, `authorization_endpoint`, `token_endpoint` and `jwks_uri`.
```
Comment (rymsha) [inline docs/index.adoc:58]:
```suggestion
The JSON result will contain the additional information: `issuer`, `authorization_endpoint`, `token_endpoint` and `jwks_uri`.
```

File: docs/index.adoc [hunk 4]
```diff
@@ -45,15 +45,17 @@ It has the following format:  `[VHost source]/_/idprovider/[ID Provider name]`.
...
 
-. The Client ID
-. The Client secret
+. The Client ID. Must be set, if log in via form is enabled.
+. The Client secret. Must be set, if log in via form is enabled.
 . The Issuer Identifier (URL)
 . The Authorization endpoint URL
 . The Token endpoint URL
+. The JWKS URI
+. The `oidcWellKnownEndpoint` (OpenID Provider Configuration URL), to obtain the values for the above fields and for `userinfoUrl`.
```
Comment (rymsha) [inline docs/index.adoc:54]:
```suggestion
. The `oidcWellKnownEndpoint` (Provider's OpenID Connect Discovery URL), to obtain the values for the above fields and for `userinfoUrl`.
```

File: docs/index.adoc [hunk 5]
```diff
@@ -179,15 +185,21 @@ idprovider.<idprovidername>.endSession.postLogoutRedirectUriKey=  (string, optio
...
 idprovider.<idprovidername>.endSession.additionalParameters.0.value=  (string, required)
 
-idprovider.<idprovidername>.clientId=  (string, required)
-idprovider.<idprovidername>.clientSecret=  (string, required)
+idprovider.<idprovidername>.clientId=  (string, optional)
+idprovider.<idprovidername>.clientSecret=  (string, optional)
 
-idprovider.<idprovidername>.mappings.displayName=  (string, required)
-idprovider.<idprovidername>.mappings.email=  (string, required)
+idprovider.<idprovidername>.mappings.displayName=  (string, required, default to ${userinfo.preferred_username})
```
Comment (rymsha) [inline docs/index.adoc:191]:
```suggestion
idprovider.<idprovidername>.mappings.displayName=  (string, required, defaults to ${userinfo.preferred_username})
```

File: docs/index.adoc [hunk 6]
```diff
@@ -179,15 +185,21 @@ idprovider.<idprovidername>.endSession.postLogoutRedirectUriKey=  (string, optio
...
+idprovider.<idprovidername>.mappings.email=  (string, required, default to ${userinfo.email})
```
Comment (rymsha) [inline docs/index.adoc:192]:
```suggestion
idprovider.<idprovidername>.mappings.email=  (string, required, defaults to ${userinfo.email})
```

File: docs/index.adoc [hunk 7]
```diff
@@ -196,12 +208,19 @@ idprovider.<idprovidername>.rules.forceEmailVerification=  (true | false, requir
 ==== Fields in the config
 
 * `autoinit`: Automatic initialization. If the config file contains `autoinit=true`, then during startup this app will automatically create ID providers for all settings declared in the file, if they don't already exist. For example, `idprovider.myfirstidp.someKey=someValue` and `idprovider.anotheridp.anotherKey=anotherValue` will declare two idproviders named `myfirstidp` and `anotheridp`.
+* `displayName`: Display name of ID Provider in User app.
```
Comment (rymsha) [inline docs/index.adoc:211]:
```suggestion
* `displayName`: Display name of ID Provider in Users app.
```

File: docs/index.adoc [hunk 8]
```diff
@@ -196,12 +208,19 @@ idprovider.<idprovidername>.rules.forceEmailVerification=  (true | false, requir
...
+* `description`: Description of ID Provider in User app.
```
Comment (rymsha) [inline docs/index.adoc:212]:
```suggestion
* `description`: Description of ID Provider in Users app.
```

File: docs/index.adoc [hunk 9]
```diff
@@ -234,6 +253,17 @@ idprovider.<idprovidername>.rules.forceEmailVerification=  (true | false, requir
...
+
+**Auto Login**
+
+* `autoLogin...` (`idprovider.<idprovidername>.autoLogin...` namespace): Auto login configuration. To enable auto login, the `jwksUri` config must be set.
+  ** `createUser`: Create a user. If set to `true`, a user will be created automatically if they don't exist.
+  ** `createSession`: Create a session. If set to `true`, a session will be established upon the user's login. By default, the user will be logged in with `REQUEST` scope.
+  ** `wsHeader`: Look for a token in the Sec-WebSocket-Protocol header. If set to `true`, the app will look for a token in the Sec-WebSocket-Protocol header.
+  ** `allowedAudience`: Allowed audience. A list of space-separated strings. If set, the app will only accept tokens with an audience that matches one of the values in this list.
+
+To disable the ability to log in via the form, `clientId` and `clientSecret` must be omitted.
```
Comment (rymsha) [inline docs/index.adoc:265]:
```suggestion
To disable the ability to auth by Authorization Code Flow, `clientId` should be omitted.
```

File: docs/index.adoc [hunk 10]
```diff
@@ -283,6 +313,16 @@ Example: **Difi/ID-porten/VER1**
 {zwsp} +
 
 == Upgrade Notes
+
+=== Version 3
+
+Version `3.0` of this app introduces a breaking change:
+
+- To verify the token signature the `jwksUri` field is now required in the ID Provider configuration.
+- The `clientId` and `clientSecret` fields are now optional in the ID Provider configuration. If you want to disable the ability to log in via the form, these fields must be omitted.
```
Comment (rymsha) [inline docs/index.adoc:322]:
```suggestion
- The `clientId` and `clientSecret` fields are now optional in the ID Provider configuration. If you want to disable the ability to auth by Authorization Code Flow, `clientId` should be omitted.
```

**Review by rymsha — CHANGES_REQUESTED** [review]

File: docs/index.adoc (hunk 1 above)
Comment (anatol-sialitski) [reply docs/index.adoc]: done

**Pull request closed without merging**